### ISSUE TITLE
Accumulate nodes with and without :retract-fn given

### ISSUE THREAD

**Issue by mrrodriguez** [rbrush/clara-rules#86]
I have been looking into the AccumulateNode of clara.rules.engine a bit recently to fully understand its semantics.
The one I want to ask about here centers on the :retract-fn given or not given in an Accumulator for Clara.

If everything works here as expected, then there is no real "issue".  If not, then some work could come out of this.

This is easiest to just demonstrate with a test case, mostly ripped from an existing clara.test-rules test.  I modified it to demonstrate my point.

```clj
(deftest test-accumulate-with-retract
  (let [collect-with-retract (accumulate
                              :initial-value []
                              :reduce-fn conj
                              :combine-fn concat

                              ;; Retract by removing the retracted item.
                              ;; In general, this would need to remove
                              ;; only the first matching item to achieve expected semantics.
                              :retract-fn (fn [reduced item]
                                            (remove #{item} reduced))

                              ;; Sort here and return the smallest.
                              :convert-return-fn (fn [reduced]
                                                   (first
                                                    (sort #(< (:temperature %1) (:temperature %2))
                                                          reduced))))

        collect-without-retract (assoc collect-with-retract
                                  ;; Note: Using the - Retractions do nothing by default for `clara.rules/accumulate`.
                                  :retract-fn (fn [reduced retracted] reduced))
                                       
        coldest-query (dsl/parse-query [] [[?t <- collect-with-retract :from (Temperature (< temperature 20))]])
        
        results-with-retract-fn (-> (mk-session [coldest-query])
                                    (insert (->Temperature 10 "MCI"))
                                    (insert (->Temperature 17 "MCI"))
                                    (insert (->Temperature 15 "MCI"))
                                    (insert (->Temperature 80 "MCI"))
                                    (retract (->Temperature 10 "MCI"))
                                    (query coldest-query)
                                    set)

        coldest-query (dsl/parse-query [] [[?t <- collect-without-retract :from (Temperature (< temperature 20))]])

        results-without-retract-fn (-> (mk-session [coldest-query])
                                       (insert (->Temperature 10 "MCI"))
                                       (insert (->Temperature 17 "MCI"))
                                       (insert (->Temperature 15 "MCI"))
                                       (insert (->Temperature 80 "MCI"))
                                       (retract (->Temperature 10 "MCI"))
                                       (query coldest-query)
                                        set)]

    ;; The accumulator result should be
    (is (= #{{:?t (->Temperature 15 "MCI")}}
           results-with-retract-fn))

    (is (= #{{:?t (->Temperature 15 "MCI")}}
           results-without-retract-fn))))

```

The results of this were, as I expected - a failure:

```clj
FAIL in (test-accumulate-with-retract) (test_rules.clj:398)
expected: (= #{{:?t (->Temperature 15 "MCI")}} results-without-retract-fn)
  actual: (not (= #{{:?t #clara.rules.testfacts.Temperature{:temperature 15, :location "MCI"}}} #{{:?t #clara.rules.testfacts.Temperature{:temperature 10, :location "MCI"}}}))
```

An accumulator with no provided :retract-fn does nothing upon retract by default.  This has the effect (from some perspectives at least) of potentially preserving "old" working state
with respect to the TMS.

The workaround for this, is obviously to specify a :retract-fn for your accumulator when you want to be able to "back out" propagations that are based on no-retracted facts.

I went in to "fix" this issue and implement a check in clara.rules.engine to use the :retract-fn if it is given, otherwise fall back to reaccumulating the state of working memory facts
with the retracted items removed.  This would require no :retract-fn to be given by default on the `clara.rules/accumulate` impl of course.

When I went to pursue this direction in the engine, I found that the AccumulateNode actually does not retain any memory of elements that it has previously been matched on.  So the ability to "re-accumulate the world" when no :retract-fn is given does not seem possible under the current design.

Once I thought about this a minute, I was led to think this may by design.  The current AccumulateNode impl allows for accumulators to "throw" away facts as they are accumulated in a
streaming sort of fashion.  I may have read or heard you mention something to this effect at some point in time.

I am mostly wanting to clarify that my thoughts on this are correct, that this is by design.
If this design decision has real utility, like the ability to release memory to already accumulated facts, etc. I'd be curious to hear some thoughts you have on that.

---

Continuing down this track.

I'll note that I do not think that AccumulateWithJoinFilterNode has these same properties as AccumulateNode with regards to the :retract-fn semantics.  AccumulateWithJoinFilterNode maintains
all candidate facts to be accumulated in its node memory.  Upon right-retract element retraction, the entire state of the candidate facts are re-accumulated to produce a new value.  This impl
does not even consider the usage of the :retract-fn.  An aside to that point, I'm skeptical that it could still get some benefit from if (see [1]).
So my observation is that AccumulateWithJoinFilterNode does not operate in the same "streaming" way that AccumulateNodes do.

---

My main purpose of this post was for my own understanding of these semantics.  I will note that I think it may be a bit more difficult to implement effective accumulators
for some workflows using AccumulateNode where I want a logically correct result value with respect to current (or "final") working memory.

I'll demonstrate with yet another test case that I find straight-forward:

```clj

(let [min-temp (dsl/parse-query [] [[?t <- (acc/min :temperature) :from [Temperature]]])

      session (-> (mk-session [min-temp])
                  (insert (->Temperature 1 "MCI"))
                  (insert (->Temperature 2 "LAX"))
                  (retract (->Temperature 1 "MCI"))
                  fire-rules)]

  (is (= #{{:?t 2}}
         (set (query session min-temp)))))


;; Results in

FAIL in <some-ns-here>
expected: (= #{{:?t 2}} (set (query session min-temp)))
  actual: (not (= #{{:?t 2}} #{{:?t 1}}))
false

```

This is using the min accumulator from clara.rules.accumulators namespace.  If I wanted an accumulator that could return the min Temperature *remaining* in working memory
at the end of changes, I'd need to maintain state of all previously seen facts via the reduced values and implement a :retract-fn that worked.  I don't think any previous facts could be ignored, since it is arbitrary which facts are removed at a later time.

An impl for this looks something like:
```clj

(defn min
  "Returns an accumulator that returns the minimum value of a given field."
  [field & {:keys [returns-fact]}]
  (let [reduce-fn (fn [[seen value] item]
                    (if (or (nil? value)
                            (< (field item) (field value)))

                      [(conj (or seen []) item) item]

                      [(conj (or seen []) item) value]))]
    (accumulate

     :initial-value nil

     :reduce-fn reduce-fn

     :combine-fn (fn [[seen1 value1] [seen2 value2]]
                   (if (or (nil? value1)
                           (< (field value2)
                              (field value1)))

                     [(into (or seen1 []) seen2) value2]

                     [(into (or seen1 []) seen2) value1]))
     
     :retract-fn (fn [[seen _] item]
                   (reduce reduce-fn nil (remove #(= item %) seen)))

     :convert-return-fn (if returns-fact
                          second
                          (comp field second)))))
```

This doesn't seem like something necessarily trivial to implement.  The reason being is that it is playing the role of managing the "node memory" for accumulator.
Perhaps this tradeoff is acceptable since it can still be written to work this way if the default design is not sufficient.

I think it is less obvious that AccumulateNode and AccumulateWithJoinFilterNode will behave differently though.

Thanks for the feedback!  Sorry it this was long-winded.

---

[1]

(note this may not be worth the trouble, just a thought)

1) accumulated results were stored in additional node memory (is this too much overhead?)
2) the filtering of candidates was done first
3) the retracted facts were retrieved from that filtered candidates
4) the accumulator had the :retract-fn ran in a `reduce` given previously-accumulated-result + retracted-facts

**Comment by rbrush** [comment]
The design of the existing min and max accumulators are indeed to support the common case without needing to hold onto the full history of inserted facts. However, whether this is the right decision is debatable...thinking back on it now I'd rather prefer correctness in the face of arbitrary retractions as opposed to support for arbitrary data streams. 

In any case, the alternative min implementation you posted is, indeed, the approach we'd take to solve it for cases that need that. Since accumulators can be arbitrary, only the accumulator itself knows what intermediate state it needs preserve to support a :retract-fn (if a :retract-fn even makes sense for the accumulator). So I think the accumulator should play the role of managing "node memory" here. The alternative is to require the full re-accumulation of changed data every time it's updated...but now we let the accumulator implementation make that decision, allowing optimizations in accumulators themselves. 

As for the AccumulateWithJoinFilterNodes, they aren't designed to behave any differently than the AccumulateNode for this case...they handle a special case of join operations that don't use simply equality-based unification. The AccumulateWithJoinFilterNode node has to maintain the full state since it doesn't know what data elements to filter out until a join occurs with other data...so it holds onto everything, then performs the join  with incoming tokens before it can complete the accumulation. This is more expensive but necessary for these semantics.

In the end, I think a case can be made for replacing the provided min and max operations with something that maintains the full history, and possibly providing some helper functions to make it easy to write accumulators in this style. If you feel like this is the way to go we can kick it around.

**Comment by mrrodriguez** [comment]
Thanks for the details on this.  I can see how there could be utility in leaving the AccumulateNode open for the streaming use-case.

I will agree that it would be helpful for some built-in Clara accumulators to offer possibly both of these behaviors.  That is a good point that some helper functions may be nice.  It probably wouldn't be too challenging to take an accumulator and decorate it to hold "node memory" and to perform a naiive `:retract-fn` that just re-accumulates based on the new state of memory.
Of course a better `:retract-fn` could be written in some cases depending on the operation of the accumulator.  However, that would be left up to the writer at that point.

**Comment by rbrush** [comment]
This was basically co-opted and fixed with issue #127. Closing because there isn't anything more to do on this.
